### PR TITLE
Initial mmsegmentation support

### PR DESCRIPTION
**Pull request by nicjac**
Initial `mmsegmentation` support. This is a work-in-progress.

Will populate this PR as we go!

- [x] Integration can be used end-to-end with **DeepLabV3** and **SegFormer**
- [x] W&B integration for semantic segmentation
- [x] New checkpoint export/import support (this one might have to wait as feature was added after the fork)
- [x] Custom CFG like mmdet implementation (might already be done, but should be tested)
- [x] Properly implement "hack" to avoid Pytorch doing parallel training by default in utils.py
- [x] Interpretation loop (currently using `mmdet`'s)
- [x] Updated semantic segmentation notebook to showcase `mmsegmentation` models

To figure out:
- [x] Splitter / Trainable Params / param_groups
- [x] ~~Could we use fastai's Dice and other metrics (if fastai is a hard dependency)?~~ Will keep current implementation

To add in an upcoming PR:
- [ ] Lightning integration
- [ ] CPU support
- [ ] Tests for `mmsegmentation` models

### CODE REVIEW

**Comment by lgvaz** [comment]
Heeey nicjac! This is a big and a very very helpful PR, we'll try to assist you as much as we can, feel free to ask for help at any point =)

**Comment by FraPochetti** [comment]
Hi @nicjac any update here?
cc @ai-fast-track @fstroth @potipot

**Comment by nicjac** [comment]
> Hi @nicjac any update here? cc @ai-fast-track @fstroth @potipot

Yep! We've been using this implementation internally for a while, but got sidetracked by other priorities.

Work has resumed, will update the PR and make sure it can be merged.

**Comment by nicjac** [comment]
This looks good to go to me!

File: icevision/models/mmseg/lightning/model_adapter.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        preds = self.convert_raw_predictions(
+            batch=data, raw_preds=raw_preds, records=records
+        )
+        self.accumulate_metrics(preds)
+
+        for k, v in outputs["log_vars"].items():
+            self.log(f"valid/{k}", v)
+
+        # TODO: is train and eval model automatically set by lighnting?
+        self.model.train()
```
Comment (FraPochetti) [inline icevision/models/mmseg/lightning/model_adapter.py:64]: @potipot this is not needed right?
As @nicjac says PL should handle the `.train()`/`.eval()` switch

File: icevision/models/mmseg/lightning/model_adapter.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    # Returns
+        A `LightningModule`.
+    """
+
+    def __init__(self, model: nn.Module, metrics: List[Metric] = None):
+        super().__init__(metrics=metrics)
+        self.model = model
+
+    @abstractmethod
+    def convert_raw_predictions(self, batch, raw_preds, records):
```
Comment (FraPochetti) [inline icevision/models/mmseg/lightning/model_adapter.py:30]: pure python question. How can this work without a `return` or `pass` statement inside the function?

Comment (FraPochetti) [reply icevision/models/mmseg/lightning/model_adapter.py:30]: What function gets invoked [here](https://github.com/airctic/icevision/pull/934/files#diff-b52b136bda8d73b70540f7af32ce576d54b6644eb6c25ba4f08ee46df367a760R55) exactly?

**Comment by FraPochetti** [comment]
I cannot preview the `notebooks/getting_started_semantic_segmentation.ipynb` notebook.
Is [this](https://colab.research.google.com/github/Orbis-International/icevision/blob/mmseg-initial-support/notebooks/getting_started_semantic_segmentation.ipynb) how it looks like?

File: icevision/models/mmseg/models/deeplabv3plus/backbones/resnet.py [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+base_config_path = mmseg_configs_path / f"{model_name}"
+base_weights_url = "https://download.openmmlab.com/mmsegmentation/v0.5/deeplabv3plus"
+
+resnet101_d16_mg124 = MMSegDeeplabPlusBackboneConfig(
+    backbone_type="R101-D16-MG124",
+    pre_trained_variants=[
+        {
+            "config_path": "deeplabv3plus/deeplabv3plus_r101-d16-mg124_512x1024_40k_cityscapes.py",
+            "crop_size": (512, 1024),
+            "lr_schd": "40k",
```
Comment (FraPochetti) [inline icevision/models/mmseg/models/deeplabv3plus/backbones/resnet.py:31]: what is `lr_schd`?

File: icevision/metrics/segmentation_accuracy/segmentation_accuracy.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+    # Arguments
+        ignore_class: class to ignore during metric computation (e.g. void class).
+
+    """
+
+    _seg_masks_gt = None
+    _seg_masks_pred = None
+
+    def __init__(self, ignore_class=None):
```
Comment (FraPochetti) [inline icevision/metrics/segmentation_accuracy/segmentation_accuracy.py:22]: Why don't we set this to `Void` as default?

**Comment by FraPochetti** [comment]
Out of curiosity, how do you handle the end2end inference flow with masks (e.g. return the mask on the original image, not the transformed one)?
The current implementation of `end2end_detect` ([here](https://github.com/airctic/icevision/blob/4f68a87baa6651766f406b7cc9bf8d377dc98d20/icevision/models/inference.py#L18)) supports only `bboxes`.

File: icevision/models/mmseg/README.md [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+# How the integration works
+
+The integration is very similar to that of `mmdet` (as both libraries share much of their structure).
+
+There are multiple layers of integrations, which are described here to facilitate future developments. 
+
+## Callbacks
+
+The [`MMSegmentationCallback`](icevision/models/mmseg/fastai/callbacks.py) is responsible for making the `mmsegmentation` models compatible with the fastai training loop. It ensures that:
+- The `after_create` hook is ued to wrap the model in a way that `mmsegmentation`'s conventions are translated to fit with fastai's model operations
```
Comment (FraPochetti) [inline icevision/models/mmseg/README.md:14]: Small typo.
`ued` instead of `used`

File: icevision/core/mask.py [hunk 6]
```diff
@@ -64,10 +64,11 @@ class MaskArray(Mask):
 
     # Arguments
         data: Mask array, with the dimensions: (num_instances, height, width)
+        pad_dim: bool
     """
 
-    def __init__(self, data: np.uint8):
-        if len(data.shape) == 2:
+    def __init__(self, data: np.uint8, pad_dim: bool = True):
+        if pad_dim and (len(data.shape) == 2):
```
Comment (FraPochetti) [inline icevision/core/mask.py:71]: What is `pad_dim` for?

File: icevision/models/fastai/unet/fastai/callbacks.py [hunk 7]
```diff
@@ -10,3 +11,15 @@ def before_batch(self):
         self.learn.records = self.yb[0]
         self.learn.yb = (self.xb[0][1],)
         self.learn.xb = (self.xb[0][0],)
+
+    def after_pred(self):
```
Comment (FraPochetti) [inline icevision/models/fastai/unet/fastai/callbacks.py:15]: Why was this one added?
Was there something broken with the UNet implementation?

File: icevision/models/fastai/unet/prediction.py [hunk 8]
```diff
@@ -39,10 +39,21 @@ def convert_raw_predictions(
     keep_images: bool = False,
 ) -> List[Prediction]:
     mask_preds = raw_preds.argmax(dim=1)
-    tensor_images, *_ = batch
```
Comment (FraPochetti) [inline icevision/models/fastai/unet/prediction.py:42]: Same comment as the previous file.
UNet is outside of `mmseg`.
Did anything break because of the changes to `MaskArray`, hence you were obliged to edit those files?

File: icevision/models/fastai/unet/prediction.py [hunk 9]
```diff
@@ -39,10 +39,21 @@ def convert_raw_predictions(
...
+
+    if len(batch) > 1:
+        tensor_images, tensor_gts = batch
+    else:
```
Comment (FraPochetti) [inline icevision/models/fastai/unet/prediction.py:45]: Does this happen at inference?
E.g. when there are no masks?

File: icevision/models/interpretation.py [hunk 10]
```diff
@@ -223,7 +224,7 @@ def plot_top_losses(
         samples = add_annotations(samples)
 
         dl = self.infer_dl(dataset, batch_size=batch_size)
-        preds = self.predict_from_dl(model=model, infer_dl=dl)
+        preds = self.predict_from_dl(model=model, infer_dl=dl, keep_images=True)
```
Comment (FraPochetti) [inline icevision/models/interpretation.py:227]: Why do we need `keep_images` here?
We rebuild `Prediction` objects from `sorted_samples` [here](https://github.com/airctic/icevision/pull/934/files#diff-f3f81ce939d27f5cf5ed326cf50fc5cf26d47b9d1cf92ac04a22d45c802b4807R256) which should have `img`s set properly already.

File: tests/models/mmseg/test_show_results.py [hunk 11]
```diff
@@ -0,0 +1,63 @@
+import pytest
+from icevision.all import *
+from icevision.models.mmseg.models.deeplabv3.backbones import *
+from icevision.models.mmseg.models import *
+from icevision.models.interpretation import get_samples_losses
+
```
[the remark this answers is not in the document]
Comment (nicjac) [reply tests/models/mmseg/test_show_results.py:6]: I agree! I kind of ran out of time to implement more tests. Happy to do some (either for this PR, or a future refinement pass).

File: icevision_install.sh [hunk 12]
```diff
@@ -58,6 +61,14 @@ case ${icevision_version} in
       pip install git+https://github.com/airctic/icedata.git --upgrade -q
       ;;
 
+   mmseg-integration)
+      echo "- Installing icevision from master (icevision integration)"
+      pip install git+https://github.com/Orbis-International/icevision.git@05ad41f183132275c7f31f25e484fcbff0a7cfd8#egg=icevision[all] --upgrade -q
```
Comment (FraPochetti) [inline icevision_install.sh:66]: This line needs to be edited before merging

File: icevision/models/mmseg/models/deeplabv3plus/backbones/resnet.py (hunk 3 above)
Comment (nicjac) [reply icevision/models/mmseg/models/deeplabv3plus/backbones/resnet.py:31]: It corresponds to different learning schedules used during the training of pretrained models (see https://github.com/open-mmlab/mmsegmentation/tree/master/configs/_base_/schedules).

This is one of the many complexities of `mmsegmentation` - there are many variations and permutations when it comes to pretrained models: different datasets, different learning schedules, different input size, etc...

File: icevision/metrics/segmentation_accuracy/segmentation_accuracy.py (hunk 4 above)
Comment (nicjac) [reply icevision/metrics/segmentation_accuracy/segmentation_accuracy.py:22]: This is an arbitrary choice on my part - I thought the expected default behaviour would be not to ignore any class, though I can see why `Void` would make sense here. Happy either way.

File: icevision/models/fastai/unet/fastai/callbacks.py (hunk 7 above)
Comment (nicjac) [reply icevision/models/fastai/unet/fastai/callbacks.py:15]: Indeed - the original `unet` implementation did not convert predictions to the "standard" IceVision format and was carrying out operations using the raw output format. I think this was fine before as it was kind of a stand-out model being the only way to do semantic segmentation. However, it meant that it was not compatible with everything else being built for semantic seg (in this PR and elsewhere).

For example, it was not possible to have both `unet` and `mmsegmentation` models in the same tutorial notebook as `unet` would fail with the newly introduced semantic segmentation metrics.

File: icevision/core/mask.py (hunk 6 above)
Comment (nicjac) [reply icevision/core/mask.py:71]: This change introduced the option **not to** pad the mask dimensions. Before this change, the mask array always ended up being padded.

File: icevision/models/mmseg/README.md (hunk 5 above)
Comment (nicjac) [reply icevision/models/mmseg/README.md:14]: Thanks 👍

File: icevision/metrics/segmentation_accuracy/segmentation_accuracy.py (hunk 4 above)
Comment (FraPochetti) [reply icevision/metrics/segmentation_accuracy/segmentation_accuracy.py:22]: @potipot @fstroth @ai-fast-track wdyt?

File: icevision/models/interpretation.py (hunk 10 above)
Comment (nicjac) [reply icevision/models/interpretation.py:227]: I can't double check right now but I _think_ this is related to the fact that without `keep_images` set to `True`, GT masks would not be available for subsequent operations. Happy to revisit this one if you think it is needed.

File: tests/models/mmseg/test_show_results.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (FraPochetti) [reply tests/models/mmseg/test_show_results.py:6]: I'd be more comfortable with a couple more added to this PR.
Doesn't need to be anything great.
You can totally copy the ones we already have for `mmdet`.

File: icevision/core/mask.py (hunk 6 above)
Comment (FraPochetti) [reply icevision/core/mask.py:71]: During the transforms?

File: tests/models/mmseg/test_show_results.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (FraPochetti) [reply tests/models/mmseg/test_show_results.py:6]: @ai-fast-track ?

File: icevision/models/fastai/unet/prediction.py (hunk 9 above)
Comment (nicjac) [reply icevision/models/fastai/unet/prediction.py:45]: At inference time without GT masks being available, the second condition is triggered (and `tensor_gts` is set as an array of `None`).

This is admittedly a bit hacky, but was in still in an effort to get `unet` working in the more general semantic segmentation framework.

**Comment by nicjac** [comment]
> Out of curiosity, how do you handle the end2end inference flow with masks (e.g. return the mask on the original image, not the transformed one)? The current implementation of `end2end_detect` ([here](https://github.com/airctic/icevision/blob/4f68a87baa6651766f406b7cc9bf8d377dc98d20/icevision/models/inference.py#L18)) supports only `bboxes`.

Unfortunately, it does not just yet. I have a _very_ hacky implementation for prototyping purpose but a proper implementation will be part of a refinement PR. It shouldn't take too long as we do need this for our work as well!

File: icevision/models/interpretation.py (hunk 10 above)
Comment (FraPochetti) [reply icevision/models/interpretation.py:227]: You are right that the masks would not be available in GT, given the records get unloaded.
I think this is not a problem specific to the `plot_top_losses` functionality as we get GT records from [here](https://github.com/airctic/icevision/pull/934/files/6de58ba0fcd3f1af7d9c6981522b68615fe92d8c#diff-f3f81ce939d27f5cf5ed326cf50fc5cf26d47b9d1cf92ac04a22d45c802b4807R223). We don't take them from the `preds` object. As you can see [here](https://github.com/airctic/icevision/pull/934/files/6de58ba0fcd3f1af7d9c6981522b68615fe92d8c#diff-f3f81ce939d27f5cf5ed326cf50fc5cf26d47b9d1cf92ac04a22d45c802b4807R228) from `preds` we get only the predictions. Long story short the GT records are the same we had ad inputs in the dataset, therefore they are not unloaded. They have everything set properly.

Comment (nicjac) [reply icevision/models/interpretation.py:227]: I remember adding this due to tests related to `plot_top_losses` failing - maybe it isn't actually needed?

Comment (FraPochetti) [reply icevision/models/interpretation.py:227]: I'd be surprised if it was needed. Can you check removing it and see what happens?

**Comment by FraPochetti** [comment]
> > Out of curiosity, how do you handle the end2end inference flow with masks (e.g. return the mask on the original image, not the transformed one)? The current implementation of `end2end_detect` ([here](https://github.com/airctic/icevision/blob/4f68a87baa6651766f406b7cc9bf8d377dc98d20/icevision/models/inference.py#L18)) supports only `bboxes`.
> 
> Unfortunately, it does not just yet. I have a _very_ hacky implementation for prototyping purpose but a proper implementation will be part of a refinement PR. It shouldn't take too long as we do need this for our work as well!

I am very interested in this one, so I can definitely take care of it too.

File: tests/models/mmseg/test_show_results.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (ai-fast-track) [reply tests/models/mmseg/test_show_results.py:6]: @FraPochetti  @nicjac I agree we need to have at least some basic tests to make sure nothing breaks. We can add more tests in subsequent PRs.

File: icevision/models/mmseg/lightning/model_adapter.py (hunk 1 above)
Comment (potipot) [reply icevision/models/mmseg/lightning/model_adapter.py:64]: yes, every validation step is vrapped with `.train()` call by pytorch-lightning https://pytorch-lightning.readthedocs.io/en/latest/common/lightning_module.html#validation

Comment (FraPochetti) [reply icevision/models/mmseg/lightning/model_adapter.py:64]: @nicjac I guess we can remove the line, then?

File: icevision/models/mmseg/README.md (hunk 5 above)
Comment (FraPochetti) [reply icevision/models/mmseg/README.md:14]: This is not addressed yet, right?

Comment (nicjac) [reply icevision/models/mmseg/README.md]: It is now :)

File: icevision_install.sh (hunk 12 above)
Comment (nicjac) [reply icevision_install.sh:66]: Forgot this was left in, now removed

File: icevision/models/interpretation.py (hunk 10 above)
Comment (nicjac) [reply icevision/models/interpretation.py]: I removed it, and tests still work. Doesn't seem to be break anything, so will keep it like that for now!

Comment (FraPochetti) [reply icevision/models/interpretation.py]: Thanks for checking man!

File: icevision/metrics/segmentation_accuracy/segmentation_accuracy.py (hunk 4 above)
Comment (potipot) [reply icevision/metrics/segmentation_accuracy/segmentation_accuracy.py:22]: I'm not familiar with and never used `Void`. In context like this I also usually set var to None

Comment (FraPochetti) [reply icevision/metrics/segmentation_accuracy/segmentation_accuracy.py:22]: @lgvaz what are your 2 cents here?

Comment (potipot) [reply icevision/metrics/segmentation_accuracy/segmentation_accuracy.py:22]: On another note though, why is this set as class attribute rather than instance, set on init?

Comment (FraPochetti) [reply icevision/metrics/segmentation_accuracy/segmentation_accuracy.py:22]: Does it matter?

Comment (potipot) [reply icevision/metrics/segmentation_accuracy/segmentation_accuracy.py:22]: My intuition was right, take a look at this example:
image

Comment (FraPochetti) [reply icevision/metrics/segmentation_accuracy/segmentation_accuracy.py:22]: Wait, Ia m referring to `ignore_class` which is set as an instance attribute, not a class one.

Comment (potipot) [reply icevision/metrics/segmentation_accuracy/segmentation_accuracy.py:22]: Oh, sorry I totally misunderstood the issue then.

Comment (FraPochetti) [reply icevision/metrics/segmentation_accuracy/segmentation_accuracy.py:22]: Np man.
When I asked
> Why don't we set this to Void as default?

I meant 
"Why don't we set this ignore_class to Void as default?"

Comment (FraPochetti) [reply icevision/metrics/segmentation_accuracy/segmentation_accuracy.py:22]: Thanks for the discussion guys.
Thinking about it more thoroughly, I am starting to convince myself that it could be a good idea to set `ignore_class=None` as it's done now. I am not sure but I think that `Void` is CamVid specific.
